tests/test_solvers.py

```diff
@@ -0,0 +1,295 @@
+        """Create a channel containing a set of packages."""
+        # build package data
+        package_data = collections.defaultdict(dict)
+        import conda.models.records
+        for record in packages:
+            assert isinstance(record, conda.models.records.PackageRecord)
+            package_data[record.subdir][record.fn] = {
+                key: value
+                for key, value in vars(record).items()
+                if key in (
+                    'build',
+                    'build_number',
+                    'depends',
+                    'license',
+                    'md5',
+                    'name',
+                    'sha256',
+                    'size',
+                    'subdir',
+                    'timestamp',
+                    'version',
+                    'track_features',
+                    'features',
+                )
+            }
+        # write repodata
+        for subdir in subdirs:
+            subdir_path = channel_path / subdir
+            subdir_path.mkdir(parents=True, exist_ok=True)
+            subdir_path.joinpath('repodata.json').write_text(json.dumps({
+                'info': {
+                    'subdir': subdir,
+                },
+                'packages': package_data.get(subdir, {}),
+            }))
+
+    @contextlib.contextmanager
+    def env(self, key, value):
+        old = os.environ.get(key)
+        os.environ[key] = value
+        try:
+            yield
+        except:
+            if old is not None:
+                os.environ[key] = old
+
+    @contextlib.contextmanager
+    def solver(self, *, add=(), remove=(), packages=()):
+        """Create a new solver.
+
+        Roughly equivalent to ``Solver(specs_to_add=..., specs_to_remove=...)``
+        with a custom channel containing the data specified in `packages`.
+        """
+        with tempfile.TemporaryDirectory(prefix='conda-test-repo-') as tmpdir_path:
+            tmpdir = pathlib.Path(tmpdir_path)
+            channels = tuple(
+                Channel(f'file://{tmpdir / subdir}')
+                for subdir in context.subdirs
+            )
+            self._write_channel_packages(tmpdir, context.subdirs, list(packages))
+            yield self.solver_class(
+                prefix='dummy - does not exist',
+                subdirs=context.subdirs,
+                channels=channels,
+                specs_to_add=add,
+                specs_to_remove=remove,
+            )
+
+    def install(self, *specs, packages=None):
+        """Try a install transaction in channel with a set of specified packages."""
+        if not packages:
+            packages = self.index(1)
+        with self.solver(add=specs, packages=packages) as solver:
+            return solver.solve_final_state()
+
+    def index(self, num):
+        """Get the index data of the ``helpers.get_index_r_*`` helpers."""
+        get_index = getattr(helpers, f'get_index_r_{num}')
```

Comment:
I think we should figure out if we could refactor those `get_index_r_*` functions (and their related helpers), since there is a lot of repetition in those.

Reply:
Yes, this was meant to be semi-temporary. Do you want to do that work now, or maybe just add a `TODO`/`XXX` comment?

Reply:
Hm, yeah, I don't mind doing it now if it helps the code you or Jaime work on, but if it's a distraction then we should postpone to later, you're right.